### ISSUE TITLE
Move away from Avro

### ISSUE THREAD

**Issue by mcupak** [ga4gh-beacon/specification#43]
The core API is switching to Protobuf3, we'll need to consider a similar change. A different format is an option as well.

**Comment by mcupak** [comment]
Core API is now freshly on Protobuf3. It makes sense for us to migrate as well, so that we can use it as a dependency.

+1 for including this in v.0.4.0.

**Comment by skeenan** [comment]
+1

**Comment by jrambla** [comment]
We discussed the possibility of some GA4GH initiative to speed up (and
align) such migrations. Stephen have we any news/plans on that?

**Comment by skeenan** [comment]
The team at UCSC are available to help with this. This is is a longstanding offer from @diekhans @kozbo

**Comment by jrambla** [comment]
Could we arrange a presentation to the team of the suggested way to do the
migration and the caveats that we should consider? (Or a general one to the
GA4GH teams if it makes sense and we have technology to do so).

**Comment by david4096** [comment]
Hi beacon team! I've converted the Beacon Avro to protobuf grpc: [beacon](https://github.com/david4096/schemas/blob/beacon-proto/src/main/proto/ga4gh/beacon.proto) and [beacon_service](https://github.com/david4096/schemas/blob/beacon-proto/src/main/proto/ga4gh/beacon_service.proto).

You can view the resulting documentation: http://schemas.readthedocs.io/en/beacon-proto/schemas/beacon.proto.html and http://schemas.readthedocs.io/en/beacon-proto/schemas/beacon_service.proto.html.

I have been working with a plugin for generating swagger API definitions from grpc. It requires adding more specific annotations to the IDL for the HTTP methods ([see here](https://github.com/david4096/schemas/blob/annotated/src/main/proto/ga4gh/sequence_annotation_service.proto#L14)). With annotations like these, and based on the protobuf above it ought to be possible to generate beacon clients and server stubs quite easily.

**Comment by jrambla** [comment]
At the technical conference call on Nov.29th 2016, we stated that there are some open ends in the migration, especially for mandatory elements. In the meantime, we will keep updating the Afro spec.

**Comment by skeenan** [comment]
@jrambla I don't think I heard that agreed? @mcupak @mfiume I think the decision was to get clarity or a solution to the mandatory elements asap from @david4096 or @kozbo from the API team

**Comment by mcupak** [comment]
Let's work on this in parallel, i.e. work against the current Avro version until we finalize protobuf. There's no need to stall development until this gets sorted out, our spec is simple and it shouldn't be hard to migrate upcoming changes later.

**Comment by david4096** [comment]
Hey, I'm happy to help!

Our approach has been to document which fields ought to be validated in some way beyond what the protobuf provides. This hasn't posed a problem for us. Since I'm not a developer of the format I can only point to some of the justifications. The idea is that a "required" field is "required" forever and won't validate through all interfaces. Perhaps we have a proxy layer that is only meant to store and route messages. By making some field required for a version of a message, then the proxy becomes sensitive to changes in versions of the protocol, when it should just be able to route messages to a validation, caching, or storage layer. The goal is to make it easy to separate concerns and provide better migration paths. Some more reading [here](https://capnproto.org/faq.html#how-do-i-make-a-field-required-like-in-protocol-buffers) and [here](https://developers.google.com/protocol-buffers/docs/proto).

> *Required Is Forever* You should be very careful about marking fields as required. If at some point you wish to stop writing or sending a required field, it will be problematic to change the field to an optional field – old readers will consider messages without this field to be incomplete and may reject or drop them unintentionally. You should consider writing application-specific custom validation routines for your buffers instead. Some engineers at Google have come to the conclusion that using required does more harm than good; they prefer to use only optional and repeated. However, this view is not universal.

In general, I've found it easier to work with messages where everything is optional, however, providing better validation layers is an ongoing [goal](https://github.com/biosustain/protov). I would welcome some more general protocol buffers validation. If it is a blocking requirement that automated tools can perform validation for required fields, you might look into using protocol buffers `options` that are passed down with the descriptor and can be used by applications consuming the service definition. This is in essence what we've done by using the `google.api.http` annotations in the GA4GH schemas, which allow later stages to programmatically access the HTTP routes.

Please let me know how else I can help! To some engineers it seems odd to say, everything is optional, but remember, protocol buffers is an interchange format, [not a message validation tool](https://groups.google.com/forum/#!topic/protobuf/ncE0GZbslao).

**Comment by mcupak** [comment]
Branch for further development: [develop-proto](https://github.com/ga4gh/beacon-team/tree/develop-proto)

**Comment by kozbo** [comment]
+1 as this is a prerequisite to #25

**Comment by mcupak** [comment]
Note: This issue should be closed once `develop-proto` is merged into `develop`. After that, `develop-proto` should be deleted and removed from `.travis.yml`.
